ansible/roles/test/files/ptftests/populate_fdb.py

```diff
@@ -0,0 +1,173 @@
+
+            Args:
+                None
+
+            Returns:
+                None
+        '''
+        if config["log_dir"] is not None:
+            self.dataplane.stop_pcap()
+
+    def __convertMacToInt(self, mac):
+        '''
+            Converts MAC address to integer
+
+            Args:
+                mac (str): MAC Address
+
+            Returns:
+                mac (int): integer representation of MAC address
+        '''
+        return int(mac.translate(None, ":.- "), 16)
+
+    def __convertMacToStr(self, mac):
+        '''
+            Converts MAC address to string
+
+            Args:
+                mac (int): MAC Address
+
+            Returns:
+                mac (str): string representation of MAC address
+        '''
+        mac = "{:012x}".format(mac)
+        return ":".join(mac[i : i + 2] for i in range(0, len(mac), 2))
+
+    def __prepareVmIp(self):
+        '''
+            Prepares VM IP addresses
+
+            Args:
+                None
+
+            Returns:
+                vmIp (dict): Map containing vlan to VM IP address
+        '''
+        vmIp = {}
+        for vlan, config in self.configData['vlan_interfaces'].items():
+            prefixLen = self.configData['vlan_interfaces'][vlan]['prefixlen']
+            ipCount = 2**(32 - prefixLen) - 3
+            numDistinctIp = self.packetCount * self.macToIpRatio[1] / self.macToIpRatio[0]
+            self.assertTrue(
+                ipCount >= numDistinctIp,
+                'Vlan network \'{0}\' does not support the requested number of IPs \'{1}\''.format(
+                    ipCount,
+                    numDistinctIp
+                )
+            )
+            vmIp[vlan] = ipaddress.ip_address(unicode(config['addr'])) + 1
+
+        return vmIp
+
+    def __populateDutFdb(self):
+        '''
+            Populates DUT FDB entries
+
+            Args:
+                None
+
+            Returns:
+                None
+        '''
+        packet = testutils.simple_tcp_packet(
+            eth_dst=self.dutMac,
+            tcp_sport=self.TCP_SRC_PORT,
+            tcp_dport=self.TCP_DST_PORT
+        )
+        vmIp = self.__prepareVmIp()
+        macInt = self.__convertMacToInt(self.startMac)
+        numMac = numIp = 0
+        for i in range(self.packetCount):
+            port = i % len(self.configData['vlan_ports'])
+            vlan = self.configData['vlan_ports'][port]['vlan']
+
+            if i % self.macToIpRatio[1] == 0:
+                mac = self.__convertMacToStr(macInt + i)
+                numMac += 1
+            if i % self.macToIpRatio[0] == 0:
+                vmIp[vlan] = ipaddress.ip_address(unicode(vmIp[vlan])) + 1
+                numIp += 1
+
+            packet[scapy.Ether].src = mac
+            packet[scapy.IP].src = str(vmIp[vlan])
+            packet[scapy.IP].dst = self.configData['vlan_interfaces'][vlan]['addr']
+            testutils.send(self, self.configData['vlan_ports'][port]['index'], packet)
```

Comment:
I'll defer to @yxieca because he's more familiar with FDB but I had trouble following what was going on here.

I guess the biggest hurdle for me is that there aren't any obvious asserts or if-checks in this sequence, so I'm not sure what we're checking or where we're checking it or just what the expected flow is. If you could make those checks more clear then I think this will be easier to follow.

Reply:
It a utility that populate FDB entries